### ISSUE TITLE
cql3: Move to ANTLR4 for parsing CQL grammar

### ISSUE THREAD

**Issue by ManManson** [scylladb/scylla#8064]
ANTLR4 is superior to ANTLR3 in many aspects:

1. It has more powerful parsing algorithm (adaptive `LL(*)` opposed to static `LL(*)`), which is capable of parsing even some explicitly left-recursive grammars. This can probably simplify our grammars a little bit.
2. Some JIT-like caching is utilized under-the-hood, making the parser faster as it processes more input. 
3. ANTLR3 is quite outdated and isn't supported anymore. ANTLR4 is still the most recent version up to date and there are at least some chances to receive help from antlr team and submit bugfixes.
4. Most probably, many bugs of ANTLR3 should be fixed in ANTLR4.

More on differences with ANTLR3 and other gained benefits can be found there: https://github.com/antlr/antlr4/blob/master/doc/faq/general.md

**Comment by slivne** [comment]
@avikivity / @dekimir / @psarna please comment on this - I want to see we agree we do this before it being done

**Comment by psarna** [comment]
I'm for it, mainly because of (3.) - we can already see problems in issues like https://github.com/scylladb/scylla/issues/7974 .
It would also be interesting to see some basic performance comparison after we migrate (e.g. by measuring the time spent in the parser by adding debug prints).

**Comment by avikivity** [comment]
The main advantage is that it is supported. The main risk is that the grammar will be evaluated slightly differently causing some regressions (I don't think it's a large risk). Another risk is that there will be lots of little differences, causing the implementation time to explode.

**Comment by nyh** [comment]
> ANTLR4 is superior to ANTLR3 in many aspects:
> 
>     1. It has more powerful parsing algorithm (adaptive `LL(*)` opposed to static `LL(*)`), which is capable of parsing even some explicitly left-recursive grammars. This can probably simplify our grammars a little bit.

Yes, but since all our grammars (CQL and DynamoDB syntax) are already written under the existing limitations, it is quite pointless to suddenly get these new features now. It's sad how in `alternator/expressions.g` I even wrote a comment on how easier it would have been to write this parser in Yacc, which I used already 35 (!) years ago. But now that we already wrote those parsers - with tricks needed to work around the parser's limits - it's water under the bridge.

>     2. Some JIT-like caching is utilized under-the-hood, making the parser faster as it processes more input.

I think this is an anti-feature. The grammar is known in advance, and all this optimization should have been done in advance, at compile-time, not at runtime. While an optimist may say "the parser gets faster as it processes more input", a pessimist might worry that "the parser starts horribly slow and gets faster later". A newly booted Scylla node already has a difficult time (e.g., no caches, a bunch of background work, bombardment by hints, etc.), making the query parser slower for newly booted nodes is a weird feature.

>     3. ANTLR3 is quite outdated and isn't supported anymore. ANTLR4 is still the most recent version up to date and there are at least some chances to receive help from antlr team and submit bugfixes.

This is the point I agree with.  The main question I see is how much of the parser will we need to change to make it run with v4. If the answer is almost-nothing, then by all means let's try it. If the answer is "a lot" (?) maybe we should drop ANTLR altogether and consider other options as well (I'm not up to date on which are the most fashionable options).

An important question to consider is how will we **test** the new parser. I'm not worried at all about the Alternator parser - I have extensive tests for it, and I believe if anything breaks in the parsing I'll know about it immediately. But do we have full test coverage in the CQL test parser? Maybe the answer is yes, but I don't know if it is.

>     4. Most probably, many bugs of ANTLR3 should be fixed in ANTLR4.
